ddtrace/internal/telemetry/writer.py

```diff
@@ -152,18 +152,19 @@ def add_event(self, payload, payload_type):
             self._events_queue.append(event)
 
     def enable(self):
-        # type: () -> None
+        # type: () -> bool
```

Comment:
The TelemetryBase._enabled should have 3 states (not two). 
 1) TelemetryBase._enabled is None 
      - This occurs during application startup. TelemetryWriter is not yet started but we would like to queue events/logs/metrics. These events will sit in memory until the telemetry writer is started. However there is a risk for memory leaks that needs to be addressed. In a follow up PR we should limit the total size of the telemetry queue.

 2) TelemetryBase._enabled is True
       - This set after the first trace is sent to the agent. Here the Telemetry writer is running and submitting events to the agent.
 3) TelemetryBase._enabled  is False
      - `DD_INSTRUMENTATION_TELEMETRY_ENABLED=False` or `_DD_TELEMETRY_METRICS_ENABLED=False`, the telemetry writer is stopped (or was never started) and no events should be queued or sent. Again there is a risk of memory leaks if we do not handle this case correctly.

There is also an edge case `TelemetryBase._enabled = True` but `TelemetryBase.status is Stopped`. This can occur if the telemetry worker thread is killed. This could occur in gevent and this needs to be handled.

Reply:
>    In a follow up PR we should limit the total size of the telemetry queue.

There is no problem right now. Count, Gauge and Rate metrics increment a value, If we add a Count metric for each request and hit the application 1.000.000 times, the metric will be:
```
{
                "common": True,
                "metric": "test-metric",
                "points": [[1642544540, 1.000.000]],
                "type": "count",
}
```
But it's true that we could have this queue size problem with distribution metrics. For that reason, those types of metrics are disabled in ASM while we discuss the limits.

> There is also an edge case `TelemetryBase._enabled = True` but `TelemetryBase.status is Stopped`. This can occur if the telemetry worker thread is killed. This could occur in gevent and this needs to be handled.

That is the problem we have just now and this PR try to fix :smiley:  (see this comment where I explain it: https://github.com/DataDog/dd-trace-py/pull/5655#discussion_r1177457265)